Brooks/LanguageRenderer/NewCSharpRenderer.hs

```diff
@@ -0,0 +1,561 @@
+instance UnaryOpSym CSharpCode where
+    type UnaryOp CSharpCode = Doc
+    notOp = return $ notOpDocD
+    negateOp = return $ negateOpDocD
+    sqrtOp = return $ text "Math.Sqrt"
+    absOp = return $ text "Math.Abs"
+    logOp = return $ text "Math.Log10"
+    lnOp = return $ text "Math.Log"
+    expOp = return $ text "Math.Exp"
+    sinOp = return $ text "Math.Sin"
+    cosOp = return $ text "Math.Cos"
+    tanOp = return $ text "Math.Tan"
+    asinOp = return $ text "Math.Asin"
+    acosOp = return $ text "Math.Acos"
+    atanOp = return $ text "Math.Atan"
+    floorOp = return $ text "Math.Floor"
+    ceilOp = return $ text "Math.Ceiling"
+
+instance BinaryOpSym CSharpCode where
+    type BinaryOp CSharpCode = Doc
+    equalOp = return $ equalOpDocD
+    notEqualOp = return $ notEqualOpDocD
+    greaterOp = return $ greaterOpDocD
+    greaterEqualOp = return $ greaterEqualOpDocD
+    lessOp = return $ lessOpDocD
+    lessEqualOp = return $ lessEqualOpDocD
+    plusOp = return $ plusOpDocD
+    minusOp = return $ minusOpDocD
+    multOp = return $ multOpDocD
+    divideOp = return $ divideOpDocD
+    powerOp = return $ text "Math.Pow"
+    moduloOp = return $ moduloOpDocD
+    andOp = return $ andOpDocD
+    orOp = return $ orOpDocD
+
+instance ValueSym CSharpCode where
+    type Value CSharpCode = (Doc, Maybe String)
+    litTrue = return $ (litTrueD, Just "true")
+    litFalse = return $ (litFalseD, Just "false")
+    litChar c = return $ (litCharD c, Just $ "\'" ++ [c] ++ "\'")
+    litFloat v = return $ (litFloatD v, Just $ show v)
+    litInt v = return $ (litIntD v, Just $ show v)
+    litString s = return $ (litStringD s, Just $ "\"" ++ s ++ "\"")
+
+    defaultChar = return $ (defaultCharD, Just "space character")
+    defaultFloat = return $ (defaultFloatD, Just "0.0")
+    defaultInt = return $ (defaultIntD, Just "0")
+    defaultString = return $ (defaultStringD, Just "empty string")
+    defaultBool = litFalse
+
+    ($->) = objVar
+    ($:) = enumElement
+
+    const = var
+    var n = return (varDocD n, Just n)
+    extVar l n = return $ (extVarDocD l n, Just $ l ++ "." ++ n)
+    self = return (selfDocD, Just "this")
+    arg n = liftPairFst (liftA2 argDocD (litInt n) argsList, Nothing)
+    enumElement en e = return (enumElemDocD en e, Just $ en ++ "." ++ e)
+    enumVar = var
+    objVar o v = liftPairFst (liftA2 objVarDocD o v, Just $ valName o ++ "." ++ valName v)
+    objVarSelf n = liftPairFst (liftA2 objVarDocD self (var n), Just $ "self." ++ n)
+    listVar n _ = var n
+    n `listOf` t = listVar n t
+    iterVar = var
+    
+    inputFunc = return $ (text "Console.ReadLine()", Nothing)
+    argsList = return $ (text "args", Nothing)
+
+    valName (CSC (v, s)) = case s of Nothing -> error $ "Attempt to print unprintable Value (" ++ render v ++ ")"
+                                     Just valstr -> valstr
+
+instance NumericExpression CSharpCode where
+    (#~) v = liftPairFst (liftA2 unOpDocD negateOp v, Nothing)
```

Comment:
looks like there is a definite opportunity to make soem helper functions to make the code below nicer.

Reply:
Various helper functions implemented in 237a480698f8dc661f69409273d05e7399d56bd8.